xds/src/main/java/io/grpc/xds/WeightedTargetLoadBalancer.java

```diff
@@ -0,0 +1,216 @@
+    for (String targetName : childBalancers.keySet()) {
+      if (!targets.containsKey(targetName)) {
+        childBalancers.get(targetName).shutdown();
+      }
+    }
+    childBalancers.keySet().retainAll(targets.keySet());
+    childHelpers.keySet().retainAll(targets.keySet());
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    logger.log(XdsLogLevel.WARNING, "Received name resolution error: {0}", error);
+    if (childBalancers.isEmpty()) {
+      helper.updateBalancingState(TRANSIENT_FAILURE, new ErrorPicker(error));
+    }
+    for (LoadBalancer childBalancer : childBalancers.values()) {
+      childBalancer.handleNameResolutionError(error);
+    }
+  }
+
+  @Override
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    logger.log(XdsLogLevel.INFO, "Shutdown");
+    for (LoadBalancer childBalancer : childBalancers.values()) {
+      childBalancer.shutdown();
+    }
+  }
+
+  private void updateBalancingState() {
+    List<WeightedChildPicker> childPickers = new ArrayList<>();
+
+    ConnectivityState overallState = null;
+    for (String name : targets.keySet()) {
+      ChildHelper childHelper = childHelpers.get(name);
+      ConnectivityState childState = childHelper.currentState;
+      overallState = aggregateState(overallState, childState);
+      if (READY == childState) {
+        int weight = targets.get(name).weight;
+        childPickers.add(new WeightedChildPicker(weight, childHelper.currentPicker));
+      }
+    }
+
+    SubchannelPicker picker;
+    if (childPickers.isEmpty()) {
+      if (overallState == TRANSIENT_FAILURE) {
+        picker = new ErrorPicker(Status.UNAVAILABLE); // TODO: more details in status
+      } else {
+        picker = XdsSubchannelPickers.BUFFER_PICKER;
+      }
+    } else {
+      picker = weightedPickerFactory.picker(childPickers);
+    }
+
+    if (overallState != null) {
+      helper.updateBalancingState(overallState, picker);
+    }
+  }
+
+  @Nullable
+  private ConnectivityState aggregateState(
+      @Nullable ConnectivityState overallState, ConnectivityState childState) {
+    if (overallState == null) {
+      return childState;
+    }
+    if (overallState == READY || childState == READY) {
+      return READY;
+    }
+    if (overallState == CONNECTING || childState == CONNECTING) {
+      return CONNECTING;
+    }
+    if (overallState == IDLE || childState == IDLE) {
+      return IDLE;
+    }
+    return overallState;
+  }
+
+  private final class ChildHelper extends ForwardingLoadBalancerHelper {
+    ConnectivityState currentState = CONNECTING;
+    SubchannelPicker currentPicker = BUFFER_PICKER;
+
+    @Override
+    public void updateBalancingState(ConnectivityState newState, SubchannelPicker newPicker) {
+      currentState = newState;
+      currentPicker = newPicker;
+      WeightedTargetLoadBalancer.this.updateBalancingState();
```

Comment:
nit: you don't need `WeightedTargetLoadBalancer.this` (may need to change the method name `updateBalancingState()`).